packages/mask/src/components/shared/WalletStatusBox/TransactionList.tsx

```diff
@@ -0,0 +1,188 @@
+        boxSizing: 'border-box',
+    },
+    link: {
+        fill: 'none',
+    },
+    linkIcon: {
+        fill: 'none',
+        width: 12,
+        height: 12,
+        marginLeft: theme.spacing(0.5),
+    },
+}))
+
+const statusTextColorMap: Record<TransactionStatusType, string> = {
+    [TransactionStatusType.NOT_DEPEND]: '#FFB915',
+    [TransactionStatusType.SUCCEED]: '#60DFAB',
+    [TransactionStatusType.FAILED]: '#FF5F5F',
+    [TransactionStatusType.CANCELLED]: '#FF5F5F',
+}
+const statusTextMap: Record<TransactionStatusType, string> = {
+    [TransactionStatusType.NOT_DEPEND]: 'Pending',
+    [TransactionStatusType.SUCCEED]: 'Success',
+    [TransactionStatusType.FAILED]: 'Failed',
+    [TransactionStatusType.CANCELLED]: 'Cancelled',
+}
+
+const getContractFunctionName = async (data: string | undefined) => {
+    if (!data) return null
+    const sig = data.slice(0, 10)
+    const name = await Services.Ethereum.getContractFunctionName(sig)
+    return name
+}
+
+interface TransactionProps extends GridProps {
+    chainId: ChainId
+    transaction: RecentTransaction
+    onClear?(tx: RecentTransaction): void
+}
+const Transaction: FC<TransactionProps> = ({ chainId, transaction: tx, onClear = noop, ...rest }) => {
+    const { t } = useI18N()
+    const { Utils } = useWeb3State()
+    const { classes, theme } = useStyles()
+
+    const web3 = useWeb3()
+    const { value: contractAddress } = useAsync(async () => {
+        if (tx.receipt?.contractAddress) return tx.receipt.contractAddress
+        if (tx.payload?.params?.[0].to) return tx.payload.params[0].to as string
+        const transaction = await web3.eth.getTransaction(tx.hash)
+        return transaction.to
+    }, [web3, tx])
+    const address = (contractAddress || '').toLowerCase()
+    const txData = tx.payload?.params?.[0].data as string | undefined
+    const { value: functionName } = useAsync(async () => {
+        const name = await getContractFunctionName(txData)
+        if (name === TransactionType.CREATE_RED_PACKET) return 'Create Lucky Drop'
+        return name ? name.replace(/_/g, ' ') : 'Contract Interaction'
+    }, [txData])
+
+    const handleClear = useCallback(() => {
+        onClear(tx)
+    }, [onClear, tx])
+
+    const domain = useMemo(() => getContractOwner(address), [address])
+
+    const [txStatus, setTxStatus] = useState(tx.status)
+
+    useEffect(() => {
+        return WalletMessages.events.transactionStateUpdated.on((data) => {
+            switch (data.type) {
+                case TransactionStateType.CONFIRMED:
+                    if (tx.hash === data.receipt.transactionHash) {
+                        setTxStatus(TransactionStatusType.SUCCEED)
+                    }
+                    break
+                case TransactionStateType.FAILED:
+                    if (tx.hash === data.receipt?.transactionHash) {
+                        setTxStatus(TransactionStatusType.FAILED)
+                    }
```

Comment:
```suggestion
                    if (tx.hash === data.receipt.transactionHash) {
                        setTxStatus(TransactionStatusType.FAILED)
                    }
```

Reply:
According to the type definition, `data.receipt` is nullable.

Reply:
But why line 119 didn't check this.

Reply:
If a transaction fails, we possibly can not get it's receipt, but it will be possible when it's confirmed